### PR TITLE
feat(Anti-DDoS): antiddos resource support alarm configuration

### PR DESCRIPTION
**Pull request by MrLiangSir**
**What this PR does / why we need it**:

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

```release-note
It is confirmed that the following parameters in the interface are discarded:
back_doors、bruce_force、high_privilege、remote_login、send_frequency、waf、weak_password
```

## PR Checklist

* [ ] Tests added/passed.
* [ ] Documentation updated.
* [ ] Schema updated.

## Acceptance Steps Performed

make testacc TEST='./huaweicloud/services/acceptance/antiddos' TESTARGS='-run TestAccAntiDdos_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/antiddos -v -run TestAccAntiDdos_basic -timeout 360m -parallel 4
=== RUN   TestAccAntiDdos_basic
=== PAUSE TestAccAntiDdos_basic
=== CONT  TestAccAntiDdos_basic
--- PASS: TestAccAntiDdos_basic (403.70s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/antiddos  403.741s

### CODE REVIEW

File: huaweicloud/services/acceptance/antiddos/resource_huaweicloud_antiddos_basic_test.go [hunk 1]
```diff
@@ -92,10 +94,23 @@ resource "huaweicloud_vpc_eip" "eip_1" {
...
+  topic_urn = huaweicloud_smn_topic.topic_1.id
+  endpoint  = "mailtest@gmail.com"
+  protocol  = "email"
+  remark    = "O&M"
+}
 
 resource "huaweicloud_antiddos_basic" "antiddos_1" {
   eip_id            = huaweicloud_vpc_eip.eip_1.id
   traffic_threshold = %d
+  topic_urn = huaweicloud_smn_topic.topic_1.id
```
Comment (deer-hang) [inline huaweicloud/services/acceptance/antiddos/resource_huaweicloud_antiddos_basic_test.go]: Field alignment

File: huaweicloud/services/antiddos/resource_huaweicloud_antiddos_basic.go [hunk 2]
```diff
@@ -80,6 +83,11 @@ func ResourceCloudNativeAntiDdos() *schema.Resource {
 				Required: true,
 				ForceNew: true,
 			},
+			"topic_urn": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
```
Comment (deer-hang) [inline huaweicloud/services/antiddos/resource_huaweicloud_antiddos_basic.go]: This field can be set to empty and does not need to be set to `Computed`.

File: huaweicloud/services/antiddos/resource_huaweicloud_antiddos_basic.go [hunk 3]
```diff
@@ -148,17 +161,47 @@ func resourceCloudNativeAntiDdosUpdate(ctx context.Context, d *schema.ResourceDa
 		return diag.FromErr(err)
 	}
 
+	topicUrn := d.Get("topic_urn").(string)
+	displayName := getDisplayName(topicUrn)
+	if topicUrn != "" && displayName != "" {
```
Comment (deer-hang) [inline huaweicloud/services/antiddos/resource_huaweicloud_antiddos_basic.go]: Missing scenario where topic_urn is equal to empty.

File: huaweicloud/services/antiddos/resource_huaweicloud_antiddos_basic.go [hunk 4]
```diff
@@ -174,22 +217,30 @@ func resourceCloudNativeAntiDdosRead(_ context.Context, d *schema.ResourceData,
...
 
 	ddosStatus := results[0]
-	log.Printf("[DEBUG] Retrieved cloud native AntiDdos %s: %#v", d.Id(), ddosStatus)
+	log.Printf("[DEBUG] Retrieved cloud native AntiDDoS %s: %#v", d.Id(), ddosStatus)
+
+	// query alarm config
+	alarmResult, err := warnalertsdk.GetWarnAlert(warnAlertClient).Extract()
+	log.Printf("[DEBUG] error query alarm configuration: %s", alarmResult.TopicUrn)
+	if err != nil {
+		return common.CheckDeletedDiag(d, err, "error retrieving AntiDDoS alarm configuration")
```
Comment (deer-hang) [inline huaweicloud/services/antiddos/resource_huaweicloud_antiddos_basic.go]: I think there is no need to perform `checkDeleted` verification here.

**Comment by MrLiangSir** [comment]
make testacc TEST='./huaweicloud/services/acceptance/antiddos' TESTARGS='-run TestAccAntiDdos_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/antiddos -v -run TestAccAntiDdos_basic -timeout 360m -parallel 4
=== RUN   TestAccAntiDdos_basic
=== PAUSE TestAccAntiDdos_basic
=== CONT  TestAccAntiDdos_basic
--- PASS: TestAccAntiDdos_basic (249.85s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/antiddos  249.895s

File: huaweicloud/services/antiddos/resource_huaweicloud_antiddos_basic.go [hunk 5]
```diff
@@ -198,12 +275,19 @@ func resourceCloudNativeAntiDdosRead(_ context.Context, d *schema.ResourceData,
 	return nil
 }
 
+func flattenTopicUrn(warnalert *warnalertsdk.WarnAlertResponse) string {
+	if warnalert.WarnConfig.AntiDDoS {
```
Comment (deer-hang) [inline huaweicloud/services/antiddos/resource_huaweicloud_antiddos_basic.go:281]: Add a pre check `if warnalert == nil {return ""}`

**Comment by MrLiangSir** [comment]
make testacc TEST='./huaweicloud/services/acceptance/antiddos' TESTARGS='-run TestAccAntiDdos_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/antiddos -v -run TestAccAntiDdos_basic -timeout 360m -parallel 4
=== RUN   TestAccAntiDdos_basic
=== PAUSE TestAccAntiDdos_basic
=== CONT  TestAccAntiDdos_basic
--- PASS: TestAccAntiDdos_basic (257.53s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/antiddos  257.577s

File: huaweicloud/services/antiddos/resource_huaweicloud_antiddos_basic.go [hunk 6]
```diff
@@ -138,27 +145,90 @@ func resourceCloudNativeAntiDdosUpdate(ctx context.Context, d *schema.ResourceDa
...
+
+	if _, err := warnalertsdk.UpdateWarnAlert(client, updateOpts).Extract(); err != nil {
+		return fmt.Errorf("error updating AntiDDoS alarm configuration: %s", err)
+	}
+
+	return nil
+}
+
+// "topic_urn":"urn:smn:cn-south-1:09f960944c80f4802f85c333e0ed1d98:tf_test"
+func getDisplayName(topUrn string) string {
```
Comment (ShiChangkuo) [inline huaweicloud/services/antiddos/resource_huaweicloud_antiddos_basic.go]: it's better to update the function name
```suggestion
func getSmnDisplayName(topUrn string) string {
```

File: huaweicloud/services/antiddos/resource_huaweicloud_antiddos_basic.go [hunk 7]
```diff
@@ -138,27 +145,90 @@ func resourceCloudNativeAntiDdosUpdate(ctx context.Context, d *schema.ResourceDa
...
+	if topicUrn != "" {
+		updateOpts = warnalertsdk.UpdateOps{
+			TopicUrn: topicUrn,
+			WarnConfig: &warnalertsdk.WarnConfig{
+				EnableAntiDDoS: utils.Bool(true),
+			},
+			DisplayName: displayName,
+		}
+	} else {
+		alarmResult, err := warnalertsdk.GetWarnAlert(client).Extract()
```
Comment (ShiChangkuo) [inline huaweicloud/services/antiddos/resource_huaweicloud_antiddos_basic.go]: why do we get the name from API?

**Comment by MrLiangSir** [comment]
make testacc TEST='./huaweicloud/services/acceptance/antiddos' TESTARGS='-run TestAccAntiDdos_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/antiddos -v -run TestAccAntiDdos_basic -timeout 360m -parallel 4
=== RUN   TestAccAntiDdos_basic
=== PAUSE TestAccAntiDdos_basic
=== CONT  TestAccAntiDdos_basic
--- PASS: TestAccAntiDdos_basic (255.77s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/antiddos  255.807s

**Pull request merged**